hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs

```diff
@@ -197,76 +520,192 @@ instance HasVariables (Action Model a) where
 deriving instance Eq (Action Model a)
 deriving instance Show (Action Model a)
 
-instance RunModel Model IO where
-  perform :: Model -> Action Model a -> LookUp IO -> IO a
-  perform Model{utxoV, alreadyContested} action lookupVar = do
+-- * ============================== REAL WORLD ==========================
+
+-- | Application monad to perform model actions. Currently it only keeps a
+-- 'UTxO' which is updated whenever transactions are valid in 'performTx'.
+newtype AppM a = AppM {runAppM :: ReaderT (IORef (UTxO, Natural)) IO a}
+  deriving newtype (Functor, Applicative, Monad, MonadIO, MonadFail, MonadThrow)
+
+instance MonadReader (UTxO, Natural) AppM where
+  ask = AppM $ ask >>= liftIO . readIORef
+
+  local f action = do
+    utxo <- ask
+    r <- newIORef (f utxo)
+    AppM $ local (const r) $ runAppM action
+
+instance MonadState (UTxO, Natural) AppM where
+  get = ask
+  put utxo = AppM $ ask >>= liftIO . flip writeIORef utxo
+
+type instance Realized AppM a = a
+
+instance RunModel Model AppM where
+  perform _m action _lookupVar = do
+    case action of
+      Decrement{actor, snapshot} -> do
+        (_, v) <- get
+        tx <- newDecrementTx actor (signedSnapshot snapshot (UnsafeSnapshotVersion v))
```

Comment:
Not sure if keeping the version in the run model is a good idea. I think we are even interested in modelling how various traces of snapshot versions are working (e.g. can't decrement twice with the same version; can close with current - 1 or current version, but no other version; ...)